rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp

```diff
@@ -368,21 +372,41 @@ uint64_t SqliteStorage::get_bagfile_size() const
 
 void SqliteStorage::initialize()
 {
-  std::string create_stmt = "CREATE TABLE topics(" \
+  std::string create_stmt = "CREATE TABLE schema(" \
+    "schema_version INTEGER PRIMARY KEY," \
+    "ros_distro TEXT NOT NULL);";
+  database_->prepare_statement(create_stmt)->execute_and_reset();
+
+  create_stmt = "CREATE TABLE metadata(" \
+    "id INTEGER PRIMARY KEY," \
+    "metadata_version INTEGER NOT NULL," \
+    "metadata TEXT NOT NULL);";
```

Comment:
What are the rows of this table intended to be? What would `id INTEGER` be set to?

Since the database schema should be changed as infrequently as possible, it seems more "future proof" to have a general purpose metadata table, like:
| `key TEXT` | `value TEXT` |
| -- | -- |
| schema_version | 1 |
| ros_distro | GALACTIC |
| duration | 11250 |
| compression_format | |
| compression_mode | |

Reply:
As regards to the question about "What are the rows of this table intended to be? What would id INTEGER be set to?"
See my explanation from previous comment in item 3

> 3. According to the update_metadata API https://github.com/ros2/rosbag2/pull/1149. We will call update_metadata twice one time right after opening storage and second time just before closing bag. Since currently we don't have update API for fields in DB perhaps we will store 2 records for metadata and store another fields repeatedly looks odd.

- Hence `id` will be a sequence number for metadata records. I expect that it will be at least 2 records of them. And when we will open storage for read we are going to read the latest one.
- `metadata_version` will be needed to be able to know how to parse serialized metadata object stored in string. According to the design proposal in #1109 we are going to serialize metadata to string in the same form as we are preparing it to store in `metadata.yaml` file. But who knows, we might change this approach in future or have some specific for deserialization based on the metadata version.

Reply:
I kind of like the key->value approach as well, but I also don't care at the end of the day specifically at this level how it's stored. As long as the rosbag2-side APIs are clean, to me this is an implementation detail of this particular storage format.